crates/validator-fetcher/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+use anyhow::Context;
+use pathfinder_common::{
+    BlockHeader,
+    CallParam,
+    CallResultValue,
+    ChainId,
+    ContractAddress,
+    EntryPoint,
+};
+use pathfinder_consensus::PublicKey;
+use pathfinder_crypto::Felt;
+use pathfinder_executor::{ExecutionState, L1BlobDataAvailability};
+use pathfinder_storage::Storage;
+use thiserror::Error;
+
+/// Validator information structure matching the Cairo contract
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ValidatorInfo {
+    pub address: ContractAddress,
+    pub public_key: PublicKey,
+    pub voting_power: u64,
+}
+
+/// Error types for validator fetching operations
+#[derive(Error, Debug)]
+pub enum ValidatorFetcherError {
+    #[error("Database error: {0}")]
+    Database(#[from] anyhow::Error),
+
+    #[error("Contract call error: {0}")]
+    ContractCall(String),
+
+    #[error("Invalid validator data: {0}")]
+    InvalidValidatorData(String),
+
+    #[error("No validators found at height {height}")]
+    NoValidators { height: u64 },
+
+    #[error("Block not found")]
+    BlockNotFound,
+
+    #[error("Unsupported network: {0}")]
+    UnsupportedNetwork(String),
+}
+
+impl From<pathfinder_executor::CallError> for ValidatorFetcherError {
+    fn from(err: pathfinder_executor::CallError) -> Self {
+        use pathfinder_executor::CallError::*;
+        match err {
+            ContractNotFound => Self::ContractCall("Contract not found".to_string()),
+            InvalidMessageSelector => Self::ContractCall("Entry point not found".to_string()),
+            ContractError(error, _) => Self::ContractCall(format!("Contract error: {error}")),
+            Internal(e) => Self::Database(e),
+            Custom(e) => Self::Database(e),
+        }
+    }
+}
+
+/// Returns the validator contract address for the given network
+fn get_validator_contract_address(
+    chain_id: ChainId,
+) -> Result<ContractAddress, ValidatorFetcherError> {
+    match chain_id {
+        ChainId::SEPOLIA_TESTNET => {
+            // Sepolia testnet mock validator contract
+            parse_contract_address(
+                "0x06473b97715c7b665923bc0629c66f2a7b8a1ce144699be12bcbe1e3278f109c",
+            )
+        }
+        ChainId::MAINNET => {
+            // TODO: Add mainnet validator contract address when available
+            Err(ValidatorFetcherError::UnsupportedNetwork(
+                chain_id.as_str().to_string(),
+            ))
+        }
+        _ => Err(ValidatorFetcherError::UnsupportedNetwork(
+            chain_id.as_str().to_string(),
+        )),
+    }
+}
+
+/// Fetches validators from a Starknet contract at a specific height
+pub fn get_validators_at_height(
+    storage: &Storage,
+    chain_id: ChainId,
+    block_header: BlockHeader,
```

Comment:
```suggestion
```

This fn has access to `storage` so maybe it could accept `BlockId` instead and fetch the header from the DB internally?